server/service/src/sync/central_data_synchroniser_v6.rs

```diff
@@ -96,25 +95,30 @@ impl SynchroniserV6 {
         let cursor_controller = CursorController::new(KeyType::SyncPullCursorV6);
         // TODO protection from infinite loop
         loop {
-            let cursor = cursor_controller.get(connection)?;
+            let start_cursor = cursor_controller.get(connection)?;
 
             let SyncBatchV6 {
                 end_cursor,
                 total_records,
-                records,
                 is_last_batch,
+                records,
             } = self
                 .sync_api_v6
-                .pull(cursor, batch_size, is_initialised)
+                .pull(start_cursor, batch_size, is_initialised)
                 .await?;
 
             logger.progress(SyncStepProgress::PullCentralV6, total_records)?;
 
-            for SyncRecordV6 { cursor, record } in records {
-                let buffer_row = record.to_buffer_row(None)?;
-
-                insert_one_and_update_cursor(connection, &cursor_controller, &buffer_row, cursor)?;
-            }
+            let last_cursor_in_batch = records.last().map(|r| r.cursor).unwrap_or(start_cursor);
+            let sync_buffer_rows =
+                CommonSyncRecord::to_buffer_rows(records.into_iter().map(|r| r.record).collect())?;
+            // Upsert sync buffer rows in a transaction together with cursor update
+            connection
+                .transaction_sync(|t_con| {
+                    SyncBufferRowRepository::new(t_con).upsert_many(&sync_buffer_rows)?;
+                    cursor_controller.update(t_con, last_cursor_in_batch + 1)
+                })
+                .map_err(|e| e.to_inner_error())?;
 
             cursor_controller.update(connection, end_cursor + 1)?;
```

Comment:
Agree, although i didn't want to make any logical changes in the PR. Added a todo comment: f0439566951d4b551be9fe35db56ea46f3dfced0, since this PR is approved I think it should be sufficient to record the observation for future (no harm with existing logic apart from possible confusion)